ui-frontend/packages/catalog-ui-search/src/main/webapp/component/timeline/timeline.tsx

```diff
@@ -0,0 +1,1034 @@
+    .flatMap((o) => Object.keys(o)) //[created]
+    .uniq()
+    .value()
+}
+
+// Types
+export type TimelineItem = {
+  id: string
+  selected: boolean
+  data?: any
+  attributes: {
+    [key: string]: Moment
+  }
+}
+
+type Bucket = {
+  x1: number
+  x2: number
+  selected: boolean
+  items: TimelineItem[]
+}
+export interface TimelineProps {
+  /**
+   * Height in pixels.
+   */
+  height: number
+
+  /**
+   * Mode that the timeline should be in.
+   */
+  mode?: 'single' | 'range'
+
+  /**
+   * Timezone to use when displaying times.
+   */
+  timezone: string
+
+  /**
+   * Date format to use when displaying times.
+   */
+  format: string
+
+  /**
+   * TimelineItem points
+   */
+  data?: TimelineItem[]
+
+  /**
+   * Alias Map for date attributes
+   */
+  dateAttributeAliases?: { [key: string]: string }
+
+  /**
+   * Called when the done button is clicked, providing the current selection range.
+   */
+  onDone?: (selectionRange: Moment[]) => void
+
+  /**
+   * Called when the a selection is made.
+   */
+  onSelect?: (data: TimelineItem[]) => void
+
+  /**
+   * Render function for tooltips
+   */
+  renderTooltip?: (data: TimelineItem[]) => any
+
+  /**
+   * Height offset to combat issues with dynamic heights when rendering the timeline.
+   */
+  heightOffset?: number
+
+  /**
+   * Called when a date is copied to the clipboard.
+   */
+  onCopy?: (copiedValue: string) => void
+
+  /**
+   * Minimum date bounds to render items between.
+   */
+  min?: Moment
+
+  /**
+   * Maximum date bounds to render items between.
+   */
+  max?: Moment
+}
+
+/*
+ * TODOS
+ * --------------------
+ *
+ * 1. On hover should work when the on hover is behind the area marker while still letting you brush drag (if possible)
+ */
+
+// Please see https://alignedleft.com/tutorials/d3/scales for more information about d3 scales.
+export const Timeline = (props: TimelineProps) => {
+  /**
+   * The useRef Hook creates a variable that "holds on" to a value across rendering
+   * passes. In this case it will hold our component's SVG DOM element. It's
+   * initialized null and React will assign it later (see the return statement)
+   */
+  const rootRef = useRef(null)
+  const d3ContainerRef = useRef(null)
+  const hoverLineRef = useRef(null)
+  const hoverLineTextRef = useRef(null)
+  const leftMarkerRef = useRef(null)
+  const rightMarkerRef = useRef(null)
+  const brushBarRef = useRef(null)
+
+  const { min, max } = props
+
+  const [width, setWidth] = useState(0)
+  const height = props.height
+
+  const heightOffset = props.heightOffset ? props.heightOffset : 0
+
+  const possibleDateAttributes = getPossibleDateAttributes(props.data || [])
+
+  const timescale = getTimescaleFromWidth(width, min, max)
+
+  const [xScale, setXScale] = useState(() => timescale)
+
+  const [xAxis, setXAxis] = useState(() =>
+    d3.axisBottom(xScale).tickSize(AXIS_HEIGHT).tickFormat(multiFormat)
```

Comment:
📓 Most of this code is unchanged. I updated things to use Moment and add the tickFormat.